### PR TITLE
SONARJS-745 Support "async" functions

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/declaration/FunctionDeclarationTreeImpl.java [hunk 1]
```diff
@@ -37,39 +37,41 @@
...
+    InternalSyntaxToken functionKeyword, InternalSyntaxToken starToken,
+    IdentifierTreeImpl name, ParameterListTreeImpl parameters, BlockTreeImpl body
+  ) {
+    return new FunctionDeclarationTreeImpl(null, functionKeyword, starToken, name, parameters, body);
   }
 
-  public FunctionDeclarationTreeImpl(
-    InternalSyntaxToken functionKeyword, IdentifierTreeImpl name,
-    ParameterListTreeImpl parameters, BlockTreeImpl body
+  private FunctionDeclarationTreeImpl(
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/declaration/FunctionDeclarationTreeImpl.java]: MINOR Move this constructor to comply with Java Code Conventions. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1213)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/FunctionExpressionTreeImpl.java [hunk 2]
```diff
@@ -40,91 +38,37 @@
...
-    this.functionKeyword = functionKeyword;
-    this.star = null;
-    this.name = name;
-    this.parameters = parameters;
-    this.body = body;
-
-    this.kind = kind;
-
-    this.functionType = FunctionType.create(this);
+  public static FunctionExpressionTree createGenerator(SyntaxToken functionKeyword, SyntaxToken star, @Nullable IdentifierTree name, ParameterListTree parameters, BlockTreeImpl body) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/FunctionExpressionTreeImpl.java]: MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/FunctionExpressionTreeImpl.java [hunk 3]
```diff
@@ -40,91 +38,37 @@
...
-    this.functionKeyword = functionKeyword;
-    this.star = null;
-    this.name = null;
-    this.parameters = parameters;
-    this.body = body;
-
-    this.kind = kind;
-
-    this.functionType = FunctionType.create(this);
+  public static FunctionExpressionTree create(@Nullable SyntaxToken asyncToken, SyntaxToken functionToken, @Nullable IdentifierTree name, ParameterListTree parameters, BlockTreeImpl body) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/FunctionExpressionTreeImpl.java]: MINOR Split this 189 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 4]
```diff
@@ -1374,13 +1355,13 @@ public AccessorMethodDeclarationTree accessor(
     return new AccessorMethodDeclarationTreeImpl(staticToken.orNull(), accessorToken, name, parameters, body);
   }
 
-  public FunctionDeclarationTreeImpl functionAndGeneratorDeclaration(
-    InternalSyntaxToken functionToken, Optional<InternalSyntaxToken> starToken, IdentifierTreeImpl name, ParameterListTreeImpl parameters, BlockTreeImpl body
+  public FunctionDeclarationTree functionAndGeneratorDeclaration(
+    Optional<InternalSyntaxToken> asyncToken, InternalSyntaxToken functionToken, Optional<InternalSyntaxToken> starToken, IdentifierTreeImpl name, ParameterListTreeImpl parameters, BlockTreeImpl body
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MINOR Split this 199 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MINOR 4 minor

Watch the comments in this conversation to review them.

**Comment by pynicolas** [comment]
@vilchik-elena org.sonar.javascript.se.ExpressionStack has to be updated: currently, it throws an exception on await.

**Comment by pynicolas** [comment]
@vilchik-elena The following code fails with a parsing error but it seems to be accepted by babel:
```
async x => await y;
```

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FunctionTree.java [hunk 5]
```diff
@@ -21,9 +21,15 @@
...
 
 import java.util.List;
+import javax.annotation.Nullable;
 import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
 
 public interface FunctionTree extends Tree {
+
+  @Nullable
+  SyntaxToken asyncToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FunctionTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**